server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java

```diff
@@ -1768,7 +1832,21 @@ public String getWriteableName() {
         public void writeTo(StreamOutput out) throws IOException {
             assert after != null : "should only write instances that were diffed from this node's state";
             if (out.getTransportVersion().onOrAfter(DIFFABLE_VERSION)) {
-                mapDiff.writeTo(out);
+                if (out.getTransportVersion().before(TransportVersions.PROJECT_ID_IN_SNAPSHOT)) {
+                    DiffableUtils.jdkMapDiffWithUpdatedKeys(mapDiff, projectRepo -> {
+                        if (ProjectId.DEFAULT.equals(projectRepo.projectId()) == false) {
+                            throw new IllegalArgumentException(
+                                "Cannot write instance with non-default project id "
+                                    + projectRepo.projectId()
+                                    + " to version before "
+                                    + TransportVersions.PROJECT_ID_IN_SNAPSHOT
+                            );
```

Comment:
Nit: should this be a runtime exception or an assertion? I don't think this is something we really _can not_ do, it looks more like something we shouldn't do.

Reply:
I added an assertion on top of the exception. I kept the exception as well because I think it is something that we "cannot do", i.e. no correct way to send a non-default project to a node that does not understand multi-project snapshots. It is also something we shouldn't do which warrants an assertion.